Arrays & Vectors project - Pull Request

DvdHalliday opened the pull request:
For this wee program I created two types of templated classes according to instructions. I managed to create a unified way to handle how the user was asked to create a new shape in any of the three menus, as well as displaying them. The latter I accomplished with the use of a double pointer.

DvdHalliday committed:
Project Files

.sln and .vcsprojs

DvdHalliday committed:
Part One

Pretty straight forward, might have to add some other perks to them in the future.

DvdHalliday committed:
Part Two

Added a main class named YourShapeApp, here I created the option for every menu as well as already the main menu.

DvdHalliday committed:
Part One complement

Stylizing the cpp files while also adding the const static ShapeName for every shape

DvdHalliday committed:
Part Three

Added the source file with a main function.
Also added the BuiltIn menu, and the request enter method

DvdHalliday committed:
Part Four

Simple innit?

DvdHalliday committed:
Part Five

Added every method const & no-const

DvdHalliday committed:
Part Six

Added copy constructors for square and circle. Added every function of this part.

DvdHalliday committed:
Part 7

Workin' 'ight, mistr'

DvdHalliday committed:
Part 8

Sum' basic doin'

DvdHalliday committed:
Part Nine

PRETTY similar to the static array

DvdHalliday committed:
Part 10

Done ;33 I did a function to copy in a loop the elements. it is private.

DvdHalliday committed:
Part Eleven

Not quite sure if that's how Resize is supposed to work hahaha

DvdHalliday committed:
Part Twelve

Simplicity is not only key; it is beauty.

DvdHalliday committed:
Part Thirteen

Added three new methods, Invert, Find and FindLast.
I also added some behaviour to squares and circles with the == operator just so I could use the Find and FindLast methods on them.

DvdHalliday committed:
Part Fourteen

Done, I also modified the way I was adding asking for new shapes and displaying the whole shape list, so all three methods now use the same new functions to do this task, I found it rather clever to proceed like this.

DvdHalliday committed:
Updating vcxproj files

DvdHalliday committed:
Small fixes before final release

Added const to the operator== assignments in square and circle.
Added a comment for TDynamicArray and handled the case the array is empty in PopBack
Checked const correctness for the main class
On the main class also modified the while for a do while in the dynamic menu, looks cleaner

DvdHalliday committed:
Revert "Modifying constructors"

This reverts commit 33c3d42458fc64a59122d97c1929f3628ff71983.

camodesigns commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Circle.cpp:
```diff
@@ -0,0 +1,32 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+float FCircle::GetArea() const
+{
+	return Radius * Radius * 3.141592653589793238462643383279502884f;
```
This is fine, however I can suggest that you use #define _USE_MATH_DEFINES and use #include <cmath> so you can use M_PI instead of the entire value of pi.

camodesigns commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Shape.h:
```diff
@@ -0,0 +1,13 @@
+	const static std::string ShapeName;
+
+public:
+	virtual float GetArea() const;
```
This is fine however you can use:	              
virtual float GetArea() const = 0;
virtual float GetPerimeter() const = 0;   
  in this case to have to define these for the father (Shape), however this makes it mandatory to define it in the children, which in this case is what we are looking for, so we can make it a little clearer and we avoid defining things that will not be used more in the children (Square and circle) than in the father (Shape). I learned this from leader JD, interesting don't you think?

camodesigns reviewed (commented):
I don't have any great comments, except for a couple of recommendations. Not because it's wrong but because they can help you improve your code practices more

DvdHalliday committed:
Merge branch 'main' into ArraysVectors

DvdHalliday committed:
Fixing const correctness and inheritance

Added the =0  on the FShape functions, deleted the fshape.cpp file. Added some const * *const const* ** * stuff that works just fine

DvdHalliday committed:
Merge branch 'ArraysVectors' of https://github.com/DvdHalliday/TGTraining into ArraysVectors

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Circle.h:
```diff
@@ -0,0 +1,19 @@
+
+	virtual float GetArea() const override;
+	virtual float GetPerimeter() const override;
+	virtual const std::string GetShapeName() const override;
```
Remember that you ONLY want to return a const value if it is a pointer or a reference. Otherwise, you shouldn't be returning a const value, which is the case here returning a const string.

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Circle.cpp:
```diff
@@ -0,0 +1,30 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+float FCircle::GetPerimeter() const
+{
+	return 2 * Radius * 3.141592653589793238462643383279502884f;
```
Remember the .f here.

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Square.h:
```diff
@@ -0,0 +1,19 @@
+
+	virtual float GetArea() const override;
+	virtual float GetPerimeter() const override;
+	virtual const std::string GetShapeName() const override;
```
Same here about the string.

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/Square.cpp:
```diff
@@ -0,0 +1,30 @@
+//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+float FSquare::GetPerimeter() const
+{
+	return 4 * SideLength;
```
.f here too

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/StaticArray.h:
```diff
@@ -0,0 +1,78 @@
+	{
+		for (T& Iteration : Data)
+		{
+			Iteration = T(Copy);
```
Why use a copy cosntructor here to pass into the assignment iterator. Why not use 
```
Iteration = Copy
```
directly?

JDAlzate replied:
Also, it's not an iteration, it's an element or something like that. Iteration doesn't really describe the object.

DvdHalliday replied:
> Why use a copy cosntructor here to pass into the assignment iterator. Why not use
> 
> ```
> Iteration = Copy
> ```
> 
> directly?

Shouldn't I have to define the = operator if I do it like that? In case our object behaves like strings or something

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/StaticArray.h:
```diff
@@ -0,0 +1,78 @@
+	{
+		for (int i = 0; i < N; i++)
+		{
+			T TemporaryItem = T(Data[i]);
```
Same here, you're constructing objects unnecessarily.

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/DynamicArray.h:
```diff
@@ -0,0 +1,241 @@
+private:
+	int Size = 0;
+	int Capacity = 1;
+	T* Data = new T[Capacity];
```
I think it's better to not have this with an inline initializer, but instead initialize it in the constructor directly.

This works to leave the default constructor empty, but for example in the copy constructor, you may be having a bug.

You are first initializing your Data to be of size 1 and then you are appending the other array, so you'll have an extra element.

DvdHalliday replied:
Yup, I think I will change it... However I don't think it should have that last problem just because my Append will set the new size to the sum of the two lists sizes, not their capacities

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/DynamicArray.h:
```diff
@@ -0,0 +1,241 @@
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T* GetData()
+	{
+		return &Data[0];
```
This works nicely, but you can return Data directly. No need to change anything, just commenting in case you're not entirely aware.

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/DynamicArray.h:
```diff
@@ -0,0 +1,241 @@
+				return;
+			}
+			// If this isn't the case, we can just start by pushing our last element back, this way we won't have to deal with more Reserves, and we would've done this anyway
+			PushBack(Back());
```
Nice! Makes sense

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/DynamicArray.h:
```diff
@@ -0,0 +1,241 @@
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Resize(const int NewSize)
```
Read this function again. I'm sure you can figure out how to write this same function using only 5 lines of code (excluding white space). 👀

JDAlzate commented on c++ training/ArraysVectors/Arrays-Vectors/ArraysVectors/DynamicArray.h:
```diff
@@ -0,0 +1,241 @@
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	// Similar to last method, only this will return the last occurence and not the first one.
+	int FindLast(const T& Item) const
```
![LazyLebowskiGIF](https://user-images.githubusercontent.com/36243539/191073700-49103cab-4f7f-4168-9202-a2d0278ff74f.gif)

DvdHalliday replied:
@JDAlzate 
![capybara-relax](https://user-images.githubusercontent.com/109089386/191104425-3e469880-1f6c-4417-abc4-e19d0b2cd3b3.gif)

JDAlzate reviewed (changes requested):
Looks nice! Some minor comments there and an important one about constructors.

DvdHalliday committed:
JD's review commit

Fixing const correctness and some other problems overall
